File: 216_combination_sum_3.c
#include <assert.h>
#include <stdlib.h>
#include <stdio.h>

/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *columnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
int** combinationSum3(int k, int n, int** columnSizes, int* returnSize) {
    int nums[6] = {0};

    int num = 0;
    int sum = 0;
    int top = 0;
    while(1) {
        num++;
        nums[top++] = num;
        if(num > n && top==1)
            return NULL;
        sum += num;
        if(sum == n && top == k) {
            printf("1\n");
            num = nums[--top];
            sum -= num;
        }
        else if(top == k) {
            num = nums[--top];
            sum -= num;
            while(num == n-1) {
                num = nums[--top];
                sum -= num;
                if(top == 0 && num==n)
                    return NULL;
            }
        }
    }

}


int main() {
    combinationSum3(3, 8, NULL, NULL);
    return 0;
}

File: 216_combination_sum_3.py
class Solution(object):
    def combinationSum3(self, k, n):
        nums = []
        num = 0
        sum = 0
        result = []
        while num < n:
            num += 1
            if num >= 10 and nums:
                num = nums.pop()
                sum -= num
                continue
            else:
                sum += num
            nums.append(num);
            if sum==n and len(nums)==k:
                newNums = nums[:]
                result.append(newNums)
                num = nums.pop()
                sum -= num
                num = nums.pop()
                sum -= num
            elif len(nums) == k or sum>n:
                num = nums.pop()
                sum -= num
                if num==n:
                    num = nums.pop()
                    sum -= num
        return result

assert Solution().combinationSum3(3, 9) == [[1,2,6],[1,3,5],[2,3,4]]
